controllers/gitrepository_controller.go

```diff
@@ -87,410 +84,505 @@ func (r *GitRepositoryReconciler) SetupWithManagerAndOptions(mgr ctrl.Manager, o
+	// Mark the resource as under reconciliation
+	conditions.MarkReconciling(obj, meta.ProgressingReason, "")
 
-	// if reconciliation failed, record the failure and requeue immediately
-	if reconcileErr != nil {
-		r.event(ctx, reconciledRepository, events.EventSeverityError, reconcileErr.Error())
-		r.recordReadiness(ctx, reconciledRepository)
-		return ctrl.Result{Requeue: true}, reconcileErr
+	// Reconcile the storage data
+	if result, err := r.reconcileStorage(ctx, obj); err != nil || result.IsZero() {
+		return result, err
 	}
 
-	// emit revision change event
-	if repository.Status.Artifact == nil || reconciledRepository.Status.Artifact.Revision != repository.Status.Artifact.Revision {
-		r.event(ctx, reconciledRepository, events.EventSeverityInfo, sourcev1.GitRepositoryReadyMessage(reconciledRepository))
+	// Create temp dir for Git clone
+	tmpDir, err := os.MkdirTemp("", fmt.Sprintf("%s-%s-%s-", obj.Kind, obj.Namespace, obj.Name))
+	if err != nil {
+		r.Eventf(ctx, obj, events.EventSeverityError, sourcev1.StorageOperationFailedReason, "Failed to create temporary directory: %s", err)
+		return ctrl.Result{}, err
 	}
-	r.recordReadiness(ctx, reconciledRepository)
-
-	log.Info(fmt.Sprintf("Reconciliation finished in %s, next run in %s",
-		time.Now().Sub(start).String(),
-		repository.GetInterval().Duration.String(),
-	))
+	defer os.RemoveAll(tmpDir)
 
-	return ctrl.Result{RequeueAfter: repository.GetInterval().Duration}, nil
-}
-
-func (r *GitRepositoryReconciler) checkDependencies(repository sourcev1.GitRepository) error {
-	for _, d := range repository.Spec.Include {
-		dName := types.NamespacedName{Name: d.GitRepositoryRef.Name, Namespace: repository.Namespace}
-		var gr sourcev1.GitRepository
-		err := r.Get(context.Background(), dName, &gr)
-		if err != nil {
-			return fmt.Errorf("unable to get '%s' dependency: %w", dName, err)
-		}
+	// Reconcile the source from upstream
+	var artifact sourcev1.Artifact
+	if result, err := r.reconcileSource(ctx, obj, &artifact, tmpDir); err != nil || result.IsZero() {
+		return ctrl.Result{RequeueAfter: obj.Spec.Interval.Duration}, err
+	}
 
-		if len(gr.Status.Conditions) == 0 || gr.Generation != gr.Status.ObservedGeneration {
-			return fmt.Errorf("dependency '%s' is not ready", dName)
-		}
+	// Reconcile includes from the storage
+	var includes artifactSet
+	if result, err := r.reconcileInclude(ctx, obj, includes, tmpDir); err != nil || result.IsZero() {
+		return ctrl.Result{RequeueAfter: r.requeueDependency}, err
+	}
 
-		if !apimeta.IsStatusConditionTrue(gr.Status.Conditions, meta.ReadyCondition) {
-			return fmt.Errorf("dependency '%s' is not ready", dName)
-		}
+	// Reconcile the artifact to storage
+	if result, err := r.reconcileArtifact(ctx, obj, artifact, includes, tmpDir); err != nil || result.IsZero() {
+		return result, err
 	}
 
-	return nil
+	return ctrl.Result{RequeueAfter: obj.Spec.Interval.Duration}, nil
```

Comment:
We do this a lot, I propose we add a helper function to all source objects so this becomes:
```go
return ctrl.Result{RequeueAfter: obj.RequeueAfter()}, nil
```

Reply:
There already is `GetInterval() metav1.Duration`, but maybe we should deprecate that in favor of `RequeueAfter() time.Duration`.

Reply:
I'm in favour of deprecating `GetInterval` 👍

Reply:
The only thing I do not really like is that we then get mixed methods, e.g. `GetArtifact()` but also `RequeueAfter()`.